rfcs/20230110-kernel-reuse/README.md

```diff
@@ -0,0 +1,265 @@
+
+1. This is a longterm process as it requires modifying many implementations and,
+   as such, needs to be exposed on an as implemented basis.
+2. The general methodology applies to all kernel generators.
+3. Performance improves for benchdnn testing.
+
+## 3. Method of Kernel Distribution
+To begin with, we need to discuss how to distribute these kernels, as it places
+significant restrictions on implementations. At its core, there are two options:
+
+### Distribution Option 1: Compile Kernels at Build Time
+#### Pros:
+Users never need to compile a GPU kernel. Solves first compilation issues.
+
+#### Neutral:
+OpenCL C Compiler version is fixed at build time. As such users cannot benefit
+or see regressions from installing a more recent compiler.
+
+#### Cons:
+Increases library size
+
+Increases compilation time
+
+Unclear how many primitives for which this is feasible.
+
+More challenging to implement GPU kernels. This requires all kernels have a
+finite number of configurations with reasonable binary size. This is technically
+impossible for most primitives due to our API (in particular due to post-ops and
+data layouts), but we can implement this feature for a subset of kernels in
+common use. Because of this, we still need to maintain runtime compilation.
+
+#### Implementation:
+Because this method increases library size, compilation time, and still requires
+runtime compilation, this feature should be enabled via a compilation switch
+`ONEDNN_COMPILE_KERNELS_FOR_ARCH=<gpu_arch list>`. The default value should be
+`NONE` as most users will need at most 2 GPU architectures compiled, an
+integrated GPU and a discrete GPU. As such, for most users a significant
+fraction of the increased library size is not helpful.
+
+### Distribution Option 2: Compile Kernels Just in Time
+#### Pros:
+Smaller Library size
+
+Enables more opportunity for JIT compiling.
+
+#### Cons:
+Customer has to compile the OpenCL C kernel at least once
+
+#### Implementation:
+This is the current behavior, so no changes are required.
+
+### Recommendation:
+The recommendation is option 2, keep the current distribution behavior. To allow
+maximum flexibility, we should modify implementations so they can be distributed
+under option 1.
+
+## 4. Proposed Kernel Reuse Methodology
+There are effectively two methodologies to reuse kernels. First, we can
+let users explicitly control kernel reuse, or second reuse can be implicitly
+handled by the library, via some kernel registering mechanism.
+
+### Reuse Option 1: Users Rely on cl_cache for OpenCL C Kernels
+Allow users to control reusing OpenCL C kernels via the existing cl_cache. This
+recommendation has already been made to many customers.
+
+#### Pros:
+Allows user control.
+
+Only requires modifying OpenCL C kernels.
+
+#### Cons:
+Poor default behavior. The use of cl_cache requires intervention from our users
+to setup and manage the generated cache.
+
+No support for non-OpenCL C code generation.
```

Comment:
What do you think about implementing a oneDNN version of `cl_cache`? It's a very attractive option as we basically need to implement a wrapper over `clBuildProgram` (or something like that). It's local, complexity is limited and we get the desired reuse for kernels out-of-the-box once the primitive is updated to use less shape-dependent macros.

Reply:
Any kernel caching behavior we implement should align with the primitive cache. As such, we should not implement a persistent cache like `cl_cache` here. Instead the behavior should match #1105. I think this would be a reasonable way to implement a global kernel cache, from Option 4, for OpenCL kernels, but I think there are better implementation options. 

First, cache lookup is a (somewhat) performance sensitive operation. As such, it is better to perform this lookup as soon as possible. In my opinion, [gpu_primitive_t::create_kernels](https://github.com/oneapi-src/oneDNN/blob/d49bfbe6c2c0725902b43b3d15c72cb3303be55d/src/gpu/gpu_primitive.hpp#L140) is the location where it makes the most sense for such a wrapper. An actual implementation could use the `kernel_ctx` and some implementation id (kernel name is not sufficient as multiple kernels may be compiled) as the lookup key to accomplish this with (nearly) the same complexity.

Second, I think completely divorcing kernel cache implementation from kernel implementations is a mistake. A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources. Second, the caching mechanism has two failure modes: false positives failures and false negatives. A wrapper around `clBluildProgram` provides very good resiliency against false positive failures and, IMO, is the main advantage of that design. False positives will be caught in our normal batched testing process though, while false negative are not caught by current testing processes. Because of this asymmetry, I think it is more important to optimize the design to catch false negatives. As what counts as a false negative is dependent on kernel implementation, this enforces (some) interdependency between the kernel cache and the implementation.

Reply:
To clarify myself: when mentioning `cl_cache` I rather thought about having a single place where kernel caching is implemented, in a primitive-agnostic way.

I agree about performance conerns.

> A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources.

Do you have an example? We can start from the point where all kernels are cached, and gradually discuss and address exceptions.

> Second, the caching mechanism has two failure modes: false positives failures and false negatives. A wrapper around clBluildProgram provides very good resiliency against false positive failures and, IMO, is the main advantage of that design. False positives will be caught in our normal batched testing process though, while false negative are not caught by current testing processes. Because of this asymmetry, I think it is more important to optimize the design to catch false negatives. As what counts as a false negative is dependent on kernel implementation, this enforces (some) interdependency between the kernel cache and the implementation.

What do you mean by "false positive" and "false negative"?

> In my [prototype implementation](https://github.com/intel-innersource/libraries.performance.math.onednn/pull/4946), I created a new base class compile_params_t to address the above points. I created a new gpu_primitive_t::create_kernels interface which takes a compile_params_t structure (wrapped in a jit_function_generator_t) to enforce kernel cache opt-in. In addition, the implementation uses compile_params_t as a lookup key removing indirection to improve cache lookup performance. Finally, having a specific structure for compilation parameters will enable testing for false negatives on the compile_params_t structure.

Thanks. One thing that concerns me is that this implementation requires per-primitive updates. For the naïve/straightforward I would expect:

- To use 1) OpenCL file name(s) and 2) compilation option string as the key for the cache (cache is runtime, not persistent)
- Maybe rely on the primitive cache state if we need to take into account `ONEDNN_PRIMITIVE_CACHE_CAPACITY`
  - This also doesn't require per-primitive changes

Reply:
>>A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources.

>Do you have an example? We can start from the point where all kernels are cached, and gradually discuss and address exceptions.

This applies to essentially all OpenCL implementations except for gen9_eltwise. Currently, we are encoding enough problem specific parameters at compile time that a kernel cache provides, at best, the same resolution as the primitive cache. As one example, ref_eltwise encodes the global work size into compilation (see ref_eltwise.cpp:77). Because this is so ubiquitous, I think we should at least analyze an implementation for benefit before enabling caching.

 > What do you mean by "false positive" and "false negative"?
 
There are two main failure modes for a caching mechanism. False positive is the scenario that cache lookup succeeds, but does not find the correct implementation. False negative is the case where cache lookup fails even though the expected implementation is cached. False positive errors are generally caused by incorrect equality/hash behavior on the kernel cache's key. The false negative case are generally caused by passing unused/unnecessary compilation parameters or, as I have fixed in the past, passing uninitialized data. We should add tests to catch these failures modes, but any method of testing for unused/unnecessary parameters will, by necessity, be implementation dependent.
 
> Thanks. One thing that concerns me is that this implementation requires per-primitive updates. 

Thanks for the feedback. I completely agree on this concern, and I do address it in the prototype. The prototype just interprets the compile time parameters as a byte buffer and performs a hash/equality comparison on the byte buffer. This consolidates the kernel cache key's operation similar to your suggestion, but restricts compile time parameters as we need to initialize all data and avoid pointers members values. Currently, the only reason for the implementation specific functions is easy access to the structure size. I am open to changing this, it is simple to modify the kernel cache key's behavior. As this discussion is implementation specific, I think this discussion should continue in the the prototype PR, rather than in the RFC.

Reply:
> Because this is so ubiquitous, I think we should at least analyze an implementation for benefit before enabling caching.

I agree but I don't think this should necessarily impact the cache implementation choice.

> We should add tests to catch these failures modes, but any method of testing for unused/unnecessary parameters will, by necessity, be implementation dependent.

I agree on testing. However I see that we may have enough diagnostics without much assistance from the implementation side. Basically we can come up with a lot of test cases based on a single function to check whether the requested primitive/problem in the cache already:

```c++
clear_cache();
primitive_create(P1_params);
assert(in_cache(P2_params)); // This should hold if P1 and P2 params are similar enough <- this is implementation-specific
```

Analysis and refactoring of primitive implementations require a lot of work but if we want to test that externally we need to clearly define "classes" of similar primitives should must share the same kernel. Once this is defined, we don't care about the implementation side - it's up to tests to catch such false negatives.

Reply:
I agree with your assessment.  I am only mentioning these points to establish that any kernel cache implementation will require per kernel implementation work. We should strive to minimize that work. My takeaway is just that if a kernel cache implementation requires a few lines of change in each kernel implementation, this is not a significant problem. It will be significantly more work to bring implementations into a "testing/behavior compliant" state.

Reply:
Anyway disregarding implementation concerns, I do not have any issue with implementing a kernel cache. My main motivation against a kernel cache is that caching should only be used when necessary. In some sense, an implementation which requires a cache is a design flaw as there is always a usage scenario resulting in poor performance. As such, if we can provide kernel implementation which do not require a (dynamic) cache, that that the preferable solution.